eoe/man/man1/ipcs.1
'\"macro stdmacro
.if n .pH g1.ipcs @(#)ipcs	30.2 of 12/25/85
.nr X
.if \nX=0 .ds x} IPCS 1 "Interprocess Communication Utilities" "\&"
.if \nX=1 .ds x} IPCS 1 "Interprocess Communication Utilities"
.if \nX=2 .ds x} IPCS 1 "" "\&"
.if \nX=3 .ds x} IPCS "" "" "\&"
.TH \*(x}
.SH NAME
ipcs \- report inter-process communication facilities status
.SH SYNOPSIS
.B ipcs
[ options ]
.SH DESCRIPTION
.I ipcs\^
prints certain information about active inter-process communication facilities.
Without
.IR options ,
information is printed in short format for message queues,
shared memory,
and semaphores that are currently active in the system.
Otherwise, the information that is displayed is controlled
by the following
.IR options :
.PP
.TP 8
.B \-q
Print information about active message queues.
.TP &
.B \-m
Print information about active shared memory segments.
.TP &
.B \-s
Print information about active semaphores.
.sp
.PP
If any of the options
.BR \-q ,
.BR \-m ,
or
.B \-s
are specified,
information about only those indicated will be printed.
If none of these three are specified,
information about all three will be printed subject to these options:
.sp
.TP &
.B \-b
Print biggest allowable size information.
(Maximum number of bytes in messages on queue for message queues,
size of segments for shared memory,
and number of semaphores in each set for semaphores.)\ 
See below for meaning of columns in a listing.
.TP &
.B \-c
Print creator's login name and group name.
See below.
.TP &
.B \-o
Print information on outstanding usage.
(Number of messages on queue and total number of bytes in messages on queue for
message queues and number of processes attached to shared memory segments.)\ 
.TP &
.B \-p
Print process number information.
(Process
.SM ID
of last process to send a message and process
.SM ID
of last process to receive a message on message queues and process
.SM ID
of creating process and process
.SM ID
of last process to attach or detach on shared memory segments)\ 
See below.
.TP &
.B \-t
Print time information.
(Time of the last control operation that changed the access permissions for
all facilities.
Time of last
.I msgsnd
and last
.I msgrcv
on message queues,
last
.I shmat
and last
.I shmdt
on shared memory,
last
.IR semop (2)
on semaphores.)\ 
See below.
.TP &
.B \-a
Use all print
.IR options .
(This is a shorthand notation for
.BR \-b ,
.BR \-c ,
.BR \-o ,
.BR \-p ,
and
.BR \-t .)\ 
.TP &
.BI \-C " corefile\^"
Use the file
.I corefile\^
in place of
.BR /dev/kmem .
.TP &
.BI \-N " namelist\^"
The argument will be taken as the name of an alternate
.I namelist\^
.RB ( /unix
is the default).
.PD
.PP
The column headings and the meaning of the columns in an
.I ipcs\^
listing
are given below; the letters in parentheses indicate the
.I options
that cause the corresponding heading to appear;
.B all
means that the heading
always appears.
Note that these
.I options
only determine
what information is provided for each facility; they do
.I not\^
determine which facilities will be listed.
.ta .75i
.ne 5
.PP
.PD 0
.TP 16
.SM
.BR T\*S "	(all)"
Type of the facility:
.RS 20
.TP 6
.B q
message queue;
.TP
.B m
shared memory segment;
.TP
.B s
semaphore.
.RE
.TP
.SM
.BR ID\*S "	(all)"
The identifier for the facility entry.
.TP
.SM
.BR KEY\*S "	(all)"
The key used as an argument to
.IR msgget ,
.IR semget ,
or
.I shmget
to create the facility entry.
(Note:\ 
The key of a shared memory segment is changed to
.SM
.B IPC_PRIVATE
when the segment has been removed until all processes attached to the segment
detach it.)\ 
.TP
.SM
.BR MODE\*S "	(all)"
The facility access modes and flags:
The mode consists of 11 characters that are interpreted as follows:
.RS
.sp \n(PDu
The first two characters are:
.sp \n(PDu
.RS
.PD 0
.TP 4
.B R
if a process is waiting on a
.IR msgrcv ;
.TP
.B S
if a process is waiting on a
.IR msgsnd ;
.TP
.B D
if the associated shared memory segment has been removed.
It will disappear when the last process attached to the segment
detaches it;
.TP
.B C
if the associated shared memory segment is to be cleared when the
first attach is executed;
.TP
.B \-
if the corresponding special flag is not set.
.RE
.PD
.sp \n(PDu
The next 9 characters are interpreted as three sets of three bits each.
The first set refers to the owner's permissions;
the next to permissions of others in the user-group of the facility entry;
and the last to all others.
Within each set, the first character indicates permission to read,
the second character indicates permission to write
or alter the facility entry,
and the last character is currently unused.
.sp \n(PDu
The permissions are indicated as follows:
.sp \n(PDu
.RS
.PD 0
.TP 4
.B r
if read permission is granted;
.TP
.B w
if write permission is granted;
.TP
.B a
if alter permission is granted;
.TP
.B \-
if the indicated permission is
.I not
granted.
.RE
.RE
.TP
.SM
.BR OWNER\*S "	(all)"
The login name of the owner of the facility entry.
.TP
.SM
.BR GROUP\*S "	(all)"
The group name of the group of the owner of the facility entry.
.TP
.SM
.BR CREATOR\*S "	(a,c)"
The login name of the creator of the facility entry.
.TP
.SM
.BR CGROUP\*S "	(a,c)"
The group name of the group of the creator of the facility entry.
.TP
.SM
.BR CBYTES\*S "	(a,o)"
The number of bytes in messages currently outstanding on the associated
message queue.
.TP
.SM
.BR QNUM\*S "	(a,o)"
The number of messages currently outstanding on the associated message queue.
.TP
.SM
.BR QBYTES\*S "	(a,b)"
The maximum number of bytes allowed in messages outstanding on the associated
message queue.
.TP
.SM
.BR LSPID\*S "	(a,p)"
The process
.SM ID
of the last process to send a message to the associated queue.
.TP
.SM
.BR LRPID\*S "	(a,p)"
The process
.SM ID
of the last process to receive a message from the associated queue.
.TP
.SM
.BR STIME\*S "	(a,t)"
The time the last message was sent to the associated queue.
.TP
.SM
.BR RTIME\*S "	(a,t)"
The time the last message was received from the associated queue.
.TP
.SM
.BR CTIME\*S "	(a,t)"
The time when the associated entry was created or changed.
.TP
.SM
.BR NATTCH\*S "	(a,o)"
The number of processes attached to the associated shared memory segment.
.TP
.SM
.BR SEGSZ\*S "	(a,b)"
The size of the associated shared memory segment.
.TP
.SM
.BR CPID\*S "	(a,p)"
The process
.SM ID
of the creator of the shared memory entry.
.TP
.SM
.BR LPID\*S "	(a,p)"
The process
.SM ID
of the last process to attach or detach the shared memory segment.
.TP
.SM
.BR ATIME\*S "	(a,t)"
The time the last attach was completed to the associated shared memory
segment.
.TP
.SM
.BR DTIME\*S "	(a,t)"
The time the last detach was completed on the associated shared memory
segment.
.TP
.SM
.BR NSEMS\*S "	(a,b)"
The number of semaphores in the set associated with the semaphore entry.
.TP
.SM
.BR OTIME\*S "	(a,t)"
The time the last semaphore operation was completed on the set associated
with the semaphore entry.
.DT
.PD
.SH FILES
.ta \w'/etc/passwd   'u
/unix	system namelist
.br
/dev/kmem	memory
.br
/etc/passwd	user names
.br
/etc/group	group names
.DT
.SH SEE ALSO
msgop(2),
semop(2),
shmop(2)
.SH BUGS
Things can change while
.I ipcs\^
is running; the picture it gives is only a close
approximation to reality.
.\"	@(#)ipcs.1	6.2 of 9/2/83
.Ee
'\".so /pubs/tools/origin.att


eoe/cmd/ipc/ipcs.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/* #ident	"@(#)ipc:/ipcs.c	1.9" */
#ident "$Revision: 1.25 $"
/*
**	ipcs - IPC status
**	Examine and print certain things about message queues, semaphores,
**		and shared memory.
*/

#include	"sys/types.h"
#include	"sys/ipc.h"
#include	"sys/msg.h"
#include	"sys/sem.h"
#include	"sys/shm.h"
#include	<unistd.h>
#include	"errno.h"
#include	"nlist.h"
#include	"fcntl.h"
#include	"time.h"
#include	"grp.h"
#include	"pwd.h"
#include	"getopt.h"
#include	"stdio.h"
#include	"stdlib.h"
#include	"sys/sysmp.h"
#include	"sys/sysget.h"

#include 	"sys.s"
extern int syscall(int, ...);

#define MSG     0
#define MSGINFO 2

void hp(char, char *, struct ipc_perm *, int, int);
void tp(time_t);

static void lseeke(int, off_t, int);
static void reade(int fildes, void *buf, size_t nbyte);

void print_shmstat(void);
void print_semstat(void);

#if (_MIPS_SZLONG == 64)
#define NLIST   nlist64
#else
#define NLIST   nlist
#endif /* if 64bit */

struct NLIST nl[] = {		/* name list entries for IPC facilities */
	{"msgque", 0, 0, 0},
	{"msginfo", 0, 0, 0},
	{NULL}
};

char	chdr[] = "T         ID     KEY        MODE       OWNER    GROUP",
				/* common header format */
	chdr2[] = "  CREATOR   CGROUP",
				/* c option header format */
	*name = "/unix",	/* name list file */
	*mem = "/dev/kmem",	/* memory file */
	opts[] = "abcemopqstC:N:";/* allowable options for getopt */
extern char	*optarg;	/* arg pointer for getopt */
int		bflg,		/* biggest size:
					segsz on m; qbytes on q; nsems on s */
		cflg,		/* creator's login and group names */
		eflg,		/* shared memory extentions */
		mflg,		/* shared memory status */
		oflg,		/* outstanding data:
					nattch on m; cbytes, qnum on q */
		pflg,		/* process id's: lrpid, lspid on q;
					cpid, lpid on m */
		qflg,		/* message queue status */
		sflg,		/* semaphore status */
		tflg,		/* times: atime, ctime, dtime on m;
					ctime, rtime, stime on q;
					ctime, otime on s */
		err;		/* option error count */
int		Nflg;		/* specified -N */


main(
	int	argc,	/* arg count */
	char	**argv)	/* arg vector */
{
	register int	i,	/* loop control */
			md,	/* memory file file descriptor */
			o;	/* option flag */
	time_t		curtime;/* date in memory file */
	struct msqid_ds	*qds;	/* message queue data structure */
	struct msginfo msginfo;	/* message information structure */
	struct msqid_ds *qp;
	sgt_cookie_t ck;

	/* Go through the options and set flags. */
	while((o = getopt(argc, argv, opts)) != EOF)
		switch(o) {
		case 'a':
			bflg = cflg = oflg = pflg = tflg = 1;
			break;
		case 'b':
			bflg = 1;
			break;
		case 'c':
			cflg = 1;
			break;
		case 'C':
			mem = optarg;
			break;
		case 'e':
			eflg = 1;
			break;
		case 'm':
			mflg = 1;
			break;
		case 'N':
			name = optarg;
			Nflg++;
			break;
		case 'o':
			oflg = 1;
			break;
		case 'p':
			pflg = 1;
			break;
		case 'q':
			qflg = 1;
			break;
		case 's':
			sflg = 1;
			break;
		case 't':
			tflg = 1;
			break;
		case '?':
			err++;
			break;
		}
	if(err || (optind < argc)) {
		fprintf(stderr,
			"usage:  ipcs [-abcemopqst] [-C corefile] [-N namelist]\n");
		exit(1);
	}
	if((mflg + qflg + sflg) == 0)
		mflg = qflg = sflg = 1;

	/* Check out namelist and memory files. */
	if (Nflg) {
		NLIST(name, nl);
		if(!nl[MSG].n_value) {
			fprintf(stderr, "ipcs:  no namelist\n");
			exit(1);
		}
	}

	if(Nflg && ((md = open(mem, O_RDONLY)) < 0)) {
		fprintf(stderr, "ipcs:  no memory file\n");
		exit(1);
	}

	if (Nflg && (qflg || sflg)) {
		/* msgque is a struct msqid_ds **, so we have
		 * to step through the indirection. But only if
		 * Nflg is set - otherwise sysmp returns the
		 * correct value.
		 */
		lseeke(md, nl[MSG].n_value, 0);
		reade(md, &nl[MSG].n_value, sizeof(nl[MSG].n_value));
	}

	curtime = time(NULL);
	printf("IPC status from %s as of %s", mem, ctime(&curtime));

	/* Print Message Queue status report. */
	if(qflg) {
		if (Nflg) {
			lseeke(md, (long)nl[MSGINFO].n_value, 0);
			reade(md, &msginfo, sizeof(msginfo));
			lseeke(md, (long)nl[MSG].n_value, 0);
		}
		else {
			SGT_COOKIE_INIT(&ck);
			SGT_COOKIE_SET_KSYM(&ck, KSYM_MSGINFO);
			if (sysget(SGT_KSYM, (char *)&msginfo, sizeof(msginfo),
				SGT_READ, &ck) < 0) {
				perror("SGT_READ msginfo");
				exit(1);
			}
		}
		printf("%s%s%s%s%s%s\nMessage Queues:\n", chdr,
			cflg ? chdr2 : "",
			oflg ? " CBYTES  QNUM" : "",
			bflg ? " QBYTES" : "",
			pflg ? " LSPID LRPID" : "",
			tflg ? "   STIME    RTIME    CTIME " : "");
		qds = (struct msqid_ds *)calloc(msginfo.msgmni,
			 sizeof(struct msqid_ds));
		if (!qds) {
			fprintf(stderr, "no memory for message queue\n");
			exit(1);
		}
		if (!Nflg) {
			SGT_COOKIE_INIT(&ck);
			if (sysget(SGT_MSGQUEUE, (char *)qds, msginfo.msgmni *
					sizeof(struct msqid_ds), SGT_READ,
					&ck) < 0) {
				perror("SGT_READ msgqueue");
				exit(1);
			}
		}
		for (i = 0; i < msginfo.msgmni; i++) {
			if (Nflg) {
				lseeke(md, (long)nl[MSG].n_value +
					 (i*sizeof(int *)), 0);
				reade(md, &qp, sizeof(qp));
				if (qp == NULL) {
					continue;
				}
				lseeke(md, (off_t)qp, 0);
				reade(md, &qds[i], sizeof(struct msqid_ds));
			}
			if(!(qds[i].msg_perm.mode & IPC_ALLOC)) {
				continue;
			}
			hp('q', "SRrw-rw-rw-", &qds[i].msg_perm, i,
				 msginfo.msgmni);
			if(oflg)
				printf("%7u%6u", qds[i].msg_cbytes,
					 qds[i].msg_qnum);
			if(bflg)
				printf("%7u", qds[i].msg_qbytes);
			if(pflg)
				printf("%6u%6u", qds[i].msg_lspid,
					 qds[i].msg_lrpid);
			if(tflg) {
				tp(qds[i].msg_stime);
				tp(qds[i].msg_rtime);
				tp(qds[i].msg_ctime);
			}
			printf("\n");
		}
	}

	/* Print Shared Memory status report. */
	if (mflg) 
		print_shmstat();

	/* Print Semaphore facility status. */
	if (sflg) 
		print_semstat();

	exit(0);
	/* NOTREACHED */
}

extern int __semstatus(struct semstat *);

void
print_semstat(void)
{
	struct semstat	semstat;
	int rv;

	if (bflg || tflg || !qflg && !mflg)
		printf("%s%s%s%s\n", chdr,
			cflg ? chdr2 : "",
			bflg ? " NSEMS" : "",
			tflg ? "   OTIME    CTIME " : "");
	printf("Semaphores:\n");
	semstat.sm_id = 0;
	semstat.sm_location = -1LL;
	while (((rv = __semstatus(&semstat)) == 0) || (errno == EACCES)) {
		struct semid_ds	*semds;

		if (!rv && (errno == EACCES))
			continue;

		semds = &semstat.sm_semds;

		hp('s', "--ra-ra-ra-", &semds->sem_perm, semstat.sm_id, 0);
		if(bflg)
			printf("%6u", semds->sem_nsems);
		if(tflg) {
			tp(semds->sem_otime);
			tp(semds->sem_ctime);
		}
		printf("\n");
	}
}

extern int __shmstatus(struct shmstat *);

void
print_shmstat(void)
{
	struct shmstat	shmstat;
	int rv;

	if (oflg || bflg || tflg || !qflg || eflg)
		printf("%s%s%s%s%s%s%s\n", chdr,
			cflg ? chdr2 : "",
			oflg ? " NATTCH " : "",
			bflg ? "    SEGSZ " : "",
			pflg ? "  CPID  LPID" : "",
			tflg ? "   ATIME    DTIME    CTIME " : "",
			eflg ? " EXTENSIONS" : "");
	printf("Shared Memory:\n");
	shmstat.sh_id = 0;
	shmstat.sh_location = -1LL;
	while (((rv = __shmstatus(&shmstat)) == 0) || (errno == EACCES)) {
		struct shmid_ds	*shmds;

		shmds = &shmstat.sh_shmds;

		if (!rv && (errno == EACCES)) {
			if (shmds->shm_cpid == NOPID)
				continue;

			/* Special case: We created this segments, but we
			   didn't give ourselves read access.  MP_SHMSTAT
			   fills in just the cpid/cgid/cuid fields.  We'll
			   overwrite the uid and gid fields with cuid/cgid
			   just to make it more obvious who should do
			   something about this situation.
			 */
			shmds->shm_perm.uid = shmds->shm_perm.cuid;
			shmds->shm_perm.gid = shmds->shm_perm.cgid;
		}

		hp('m', "DCrw-rw-rw-", &shmds->shm_perm, 
		    shmstat.sh_id,
		    0);
		if(oflg)
			printf("%7u ", shmds->shm_nattch);
		if(bflg)
			printf("%9ld ", shmds->shm_segsz);
		if(pflg)
			printf("%6u%6u", shmds->shm_cpid,
					 shmds->shm_lpid);
		if(tflg) {
			tp(shmds->shm_atime);
			tp(shmds->shm_dtime);
			tp(shmds->shm_ctime);
		}
		if (eflg) {
			printf(" %c", shmds->shm_perm.mode & IPC_AUTORMID ? 'A' : ' ');
		}
		printf("\n");
	}
}

/*
**	lseeke - lseek with error exit
*/
static void
lseeke(
	int	fd,
	off_t	offset,
	int	whence)
{
#if _MIPS_SZPTR == 64
	offset &= 0x7fffffffffffffff;
#else
	offset &= 0x7fffffff;
#endif
	if (lseek(fd, offset, whence) == -1) {
		perror("ipcs:  seek error");
		exit(1);
	}
}

/*
**	reade - read with error exit
*/
static void
reade(
	int	fd,
	void	*bufaddr,
	size_t	size)
{
	if (read(fd, bufaddr, size) != size) {
		perror("ipcs:  read error");
		exit(1);
	}
}

/*
**	hp - common header print
*/
void
hp(
	char		type,	/* facility type */
	char		*modesp,/* ptr to mode replacement characters */
	struct ipc_perm	*permp,	/* ptr to permission structure */
	int		slot,	/* facility slot number */
	int		slots)	/* # of facility slots */
{
	register int		i,	/* loop control */
				j;	/* loop control */
	register struct group	*g;	/* ptr to group group entry */
	register struct passwd	*u;	/* ptr to user passwd entry */

	printf("%c%11d%s%#8.8x ", type, slot + slots * permp->seq,
		permp->key ? " " : " 0x", permp->key);
	for(i = 02000;i;modesp++, i >>= 1)
		printf("%c", (permp->mode & i) ? *modesp : '-');
	if((u = getpwuid(permp->uid)) == NULL)
		printf("%9d", permp->uid);
	else
		printf("%9.8s", u->pw_name);
	if((g = getgrgid(permp->gid)) == NULL)
		printf("%9d", permp->gid);
	else
		printf("%9.8s", g->gr_name);
	if(cflg) {
		if((u = getpwuid(permp->cuid)) == NULL)
			printf("%9d", permp->cuid);
		else
			printf("%9.8s", u->pw_name);
		if((g = getgrgid(permp->cgid)) == NULL)
			printf("%9d", permp->cgid);
		else
			printf("%9.8s", g->gr_name);
	}
}

/*
**	tp - time entry printer
*/
void
tp(
	time_t	time)	/* time to be displayed */
{
	register struct tm	*t;	/* ptr to converted time */

	if(time) {
		t = localtime(&time);
		printf(" %2d:%2.2d:%2.2d", t->tm_hour, t->tm_min, t->tm_sec);
	} else
		printf(" no-entry");
}
